File: python/yacht/yacht.py
"""
This exercise stub and the test suite contain several enumerated constants.

Since Python 2 does not have the enum module, the idiomatic way to write
enumerated constants has traditionally been a NAME assigned to an arbitrary,
but unique value. An integer is traditionally used because it’s memory
efficient.
It is a common practice to export both constants and functions that work with
those constants (ex. the constants in the os, subprocess and re modules).

You can learn more here: https://en.wikipedia.org/wiki/Enumerated_type
"""


# Score categories.
# Change the values as you see fit.
YACHT = 50
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = None
FOUR_OF_A_KIND = None
LITTLE_STRAIGHT = None
BIG_STRAIGHT = None
CHOICE = None

def score_yacht(dice):
    if len(set(dice)) == 1:
        return YACHT
    else:
        return 0

def score_ones(dice):
    return ONES * dice.count(1)

def score_twos(dice):
    return TWOS * dice.count(2)

def score_threes(dice):
    return THREES * dice.count(3)

def score_fours(dice):
    return FOURS * dice.count(4)

def score_fives(dice):
    return FIVES * dice.count(5)

def score_sixes(dice):
    return SIXES * dice.count(6)

def score(dice, category):
    if (category == YACHT):
        return score_yacht(dice)
    if (category == ONES):
        return score_ones(dice)
    if (category == TWOS):
        return score_twos(dice)
    if (category == THREES):
        return score_threes(dice)
    if (category == FOURS):
        return score_fours(dice)
    if (category == FIVES):
        return score_fives(dice)
    if (category == SIXES):
        return score_sixes(dice)

File: go/yacht/yacht.go
package yacht

import (
	"log"
	"reflect"
	"sort"
)

func Score(dice []int, category string) (score int) {
	switch category {
	case "ones":
		return 1 * countOf(dice, 1)
	case "twos":
		return 2 * countOf(dice, 2)
	case "threes":
		return 3 * countOf(dice, 3)
	case "fours":
		return 4 * countOf(dice, 4)
	case "fives":
		return 5 * countOf(dice, 5)
	case "sixes":
		return 6 * countOf(dice, 6)
	case "full house":
		return scoreFullHouse(dice)
	case "four of a kind":
		return scoreFourOfKind(dice)
	case "little straight":
		return scoreLittleStraight(dice)
	case "big straight":
		return scoreBigStraight(dice)
	case "choice":
		return sum(dice)
	case "yacht":
		return scoreYacht(dice)
	default:
		log.Fatalf("unrecognized category %v", category)
		return 0
	}
}

func scoreFullHouse(dice []int) (score int) {
	if !isFullHouse(dice) {
		return 0
	}
	return sum(dice)
}

func scoreFourOfKind(dice []int) (score int) {
	if !isFourOfKind(dice) {
		return 0
	}
	diceToOccurences := getOccurences(dice)
	for d, occurences := range diceToOccurences {
		if occurences >= 4 {
			return d * 4
		}
	}
	log.Fatalf("diceToOccurences did not contain dice with 4 occurences %v", diceToOccurences)
	return 0
}

func scoreLittleStraight(dice []int) (score int) {
	if !isLittleStraight(dice) {
		return 0
	}
	return 30
}

func scoreBigStraight(dice []int) (score int) {
	if !isBigStraight(dice) {
		return 0
	}
	return 30
}

func scoreYacht(dice []int) (score int) {
	if !isYacht(dice) {
		return 0
	}
	return 50
}

func countOf(dice []int, item int) (count int) {
	for _, d := range dice {
		if d == item {
			count++
		}
	}
	return count
}

func isFullHouse(dice []int) bool {
	diceToOccurences := getOccurences(dice)
	for _, occurences := range diceToOccurences {
		if occurences != 2 && occurences != 3 {
			return false
		}
	}
	return true
}

func isFourOfKind(dice []int) bool {
	diceToOccurences := getOccurences(dice)
	for _, occurences := range diceToOccurences {
		if occurences >= 4 {
			return true
		}
	}
	return false
}

func isLittleStraight(dice []int) bool {
	littleStraight := []int{1, 2, 3, 4, 5}
	sort.Ints(dice)
	return reflect.DeepEqual(dice, littleStraight)
}

func isBigStraight(dice []int) bool {
	bigStraight := []int{2, 3, 4, 5, 6}
	sort.Ints(dice)
	return reflect.DeepEqual(dice, bigStraight)
}

func isYacht(dice []int) bool {
	diceToOccurences := getOccurences(dice)
	for _, occurences := range diceToOccurences {
		if occurences != 5 {
			return false
		}
	}
	return true
}

func getOccurences(dice []int) (diceToOccurences map[int]int) {
	diceToOccurences = map[int]int{}
	for _, d := range dice {
		if _, ok := diceToOccurences[d]; !ok {
			diceToOccurences[d] = 0
		}
		diceToOccurences[d] += 1
	}
	return diceToOccurences
}

func sum(dice []int) (result int) {
	for _, d := range dice {
		result += d
	}
	return result
}
